### PR TITLE
Try to make `/readyz` check more reliable

### PR DESCRIPTION
**Pull request by ffuugoo**
Try to make `/readyz` check more reliable after cluster upgrade incident last week.

__TODO:__
- [x] Explicitly handle bootstrapping *new* cluster node
  - currently, the node may report "ready" right after startup, because for a short moment before *joining* the cluster, it considers itself the *only* node in the cluster, and `/readyz` instantly returns "ready" in that case

### All Submissions:

* [x] Contributions should target the `dev` branch. Did you create your branch from `dev`?
* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### New Feature Submissions:

1. [x] Does your submission pass tests?
2. [x] Have you formatted your code locally using `cargo +nightly fmt --all` command prior to submission?
3. [x] Have you checked your code using `cargo clippy --all --all-features` command?

### Changes to Core Features:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your core changes, as applicable?
* [ ] Have you successfully ran tests with your changes locally?

### CODE REVIEW

File: src/common/health.rs [hunk 1]
```diff
@@ -128,14 +128,20 @@ impl Task {
...
-            // - Check if cluster commit index *decresed* since last check
-            cluster_commit_index = cmp::min(current_cluster_commit_index, cluster_commit_index);
+            // TODO: Do we want to update `cluster_commit_index` here?
+            //
+            // I.e.:
+            // - If we *don't* update `cluster_commit_index`, then we will only wait till the node
+            //   catch up with the cluster commit index *at the moment the node has been started*
+            // - If we *do* update `cluster_commit_index`, then we will keep track of cluster
+            //   commit index updates and wait till the node *completely* catch up with the leader,
+            //   which might be hard (if not impossible) in some situations
```
Comment (ffuugoo) [inline src/common/health.rs:144]: First question: what do we want to do here?

Comment (generall) [reply src/common/health.rs:144]: > catch up with the cluster commit index *at the moment the node has been started*

this should be fine, I guess.

There is a middle-ground of doing this update only once, after first sync is complete

**Comment by ffuugoo** [comment]
Second question:

Initially, we *had to* allow `/readyz` to ignore errors while querying other nodes for the cluster commit index. (For the backward compatibility, because older version did not support required APIs for the `/readyz`.)

But now, we can assume, that any version running in the cluster supports `/readyz`, so we can be more strict with the error handling.

So, currently (with changes in this PR!), if the node queries other nodes for the cluster commit *for the first time*, and if *all* requests happened to fail, then the node would effectively ignore cluster commit check. It will continue to the next step (wait for the shards to turn green once).

So, do we want to make this initial cluster commit check more strict? E.g.:
- Do we want to *require* the node to query *at least* 1 other node? 
- Or, maybe, we want to *require* the node to query *at least* the majority of nodes?
- Maybe one of the above, but with fixed number of retries, so it will continue to the next step, if `N` attempts to query other nodes fail?

File: src/common/health.rs [hunk 2]
```diff
@@ -199,28 +220,32 @@ impl Task {
...
+        // other nodes are unavailable or they don't support `GetConsensusCommit` gRPC API).
+        //
+        // So this check is not 100% reliable and can give a false-positive result!
+        let cluster_commit_index = commit_indices
+            .into_iter()
+            .map(|resp| resp.into_inner().commit)
+            .max()
+            .unwrap_or(0);
+
+        Some(cluster_commit_index as _)
```
Comment (ffuugoo) [inline src/common/health.rs:248]: Also, some explanation for the changes.

_For some reason_, we decided to return 0 (which means "ignore cluster commit check"), if we did not get *at least* `required_commit_indices_count` successful responses... which is silly?

Unless someone somehow added a node from *another* cluster to the existing one, or manually modified node data on disk, or *manually* break the cluster in some elaborate way, getting *any* consensus commit is better than getting none.

So I removed this nonsense logic, and I always return `max` of *any* commit indices we have. We only return 0 if *all` requests failed.

File: src/common/health.rs [hunk 3]
```diff
@@ -128,14 +128,20 @@ impl Task {
...
+            // - Check if this is the only node in the cluster
+            let Some(_) = self.cluster_commit_index().await else {
                 self.set_ready();
                 return;
             };
 
-            // - Check if cluster commit index *decresed* since last check
-            cluster_commit_index = cmp::min(current_cluster_commit_index, cluster_commit_index);
+            // TODO: Do we want to update `cluster_commit_index` here?
+            //
```
Comment (ffuugoo) [inline src/common/health.rs:138]: We also did this `min` of the `cluster_commit_index` on each iteration, but Raft guarantees that it will never decrease cluster commit index.

And so, if during any of this continuous queries, *all* requests fail (e.g., because node lost connectivity for a moment) then `cluster_commit_index()` will return 0, and `min(cluster_commit_index, 0)` *is* 0. So we will instantly go to the next step of the `/readyz` check, even though we have *some* reasonable commit already.

Basically, I think this also makes no sense and makes `/readyz` check more flaky, so I've removed the `min` call.

File: src/common/health.rs (hunk 1 above)
Comment (ffuugoo) [reply src/common/health.rs:144]: Ok, then it should be fine as-is.

File: src/common/health.rs (hunk 2 above)
Comment (generall) [reply src/common/health.rs:264]: I may speculate that we did it for the compatibility with versions that didn't support sharing of the consensus version. Anyway, I don't see the reason for keeping the old logic either.

**Review by generall — APPROVED** [review]
Found myself struggling to remember how some of the signals work in this scenario, so took a liberty to propose additional comments.
